Rewrite Swift For TensorFlow random number generation.

jtkeeling opened the pull request:
We break the PRNG library into engines for generating random bytes and distributions for using engines to produce numbers that fit some statistical distribution.

We modify Tensor to use this new version and remove the RandomState class from Utility.

jtkeeling committed:
Rewrite Swift For TensorFlow random number generation.

We break the PRNG library into engines for generating random bytes and distributions for using engines to produce numbers that fit some statistical distribution.

We modify Tensor to use this new version and remove the RandomState class from Utility.

jtkeeling commented:
@swift-ci please test tensorflow

dan-zheng commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,141 @@
+//===----------------------------------------------------------------------===//
+
+public class UniformIntegerDistribution<T: BinaryInteger> {
+
```
Could you please remove the first empty line inside `class` declarations? Same below.

jtkeeling replied:
Done.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -679,16 +633,22 @@ public extension Tensor where Scalar : BinaryFloatingPoint {
+    self = Tensor(
+      handle: _TFHoistable {
+        let engine = engine ?? ARC4RandomEngine.global
+        let dist = NormalFloatingPointDistribution<Scalar>(mean: mean, standardDeviation: stddev)
```
Wrap to fit within 80 cols:
```swift
        let dist = NormalFloatingPointDistribution<Scalar>(
           mean: mean, standardDeviation: stddev
        )
```

jtkeeling replied:
Done. All formatting should be good now. I had forgotten it was 80 not 100 chars.

dan-zheng commented on test/TensorFlowRuntime/random.swift:
```diff
@@ -0,0 +1,183 @@
+
+RandomTests.test("CopyDoesntBreakOriginal") {
+  let engine1 = ARC4RandomEngine(seed: [0x1a, 0xda, 0x31, 0xd5, 0xcf, 0x68, 0x82, 0x21,
+                                       0xc1, 0x09, 0x16, 0x39, 0x08, 0xeb, 0xe5, 0x1d,
```
Please indent these three lines (beginning with `0xc1`, `0xeb`, `0x64`) with an extra space.
Same for the `CopyIsDistinctFromOriginal` test below.

jtkeeling replied:
Reformatted.

dan-zheng commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,141 @@
+    }
+  }
+
+  public convenience required init<T>(seed: T) where T : BinaryInteger {
```
Is there strong motivation for making this constructor generic, rather than using pinning a specific integer type? (Previously, the seed type was pinned as `Int32`.)

jtkeeling replied:
Not really, though as they are defined by calling the other initializer, it's no more effort to do this. I would suggest there are better choices than Int32 anyway, such as UInt64. Let me know if you really think we should pin a type, otherwise I'll leave as-is.

Azoy commented:
I'm not too involved in the tensorflow project, but is there a reason why you didn't just build on top of `SE-0202`? See: [Random.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/Random.swift). That way you can achieve things like `let randomFloat = Tensor<Float>(.random(in: 0 ..< 1, using: &arc4))` no? (Or whatever syntax you would like to introduce).

dan-zheng commented:
@Azoy Thanks for bringing up SE-202!
An important requirement for RNGs in machine learning is seedability, which is why we initially pursued our own seedable RNG implementation. (This is sort of at-odds with `arc4random()` on macOS, which is focused on security and isn't really seedable AFAIK.)

I think conforming the various RNGs in this PR to `RandomNumberGenerator` is pretty sensible.
The `Tensor` random initializers can take an optional `RandomNumberGenerator`:
```swift
init(randomUniform shape: TensorShape, generator: RandomNumberGenerator? = nil) {
  let generator = generator ?? ARC4RandomGenerator.global
  // ...
}
```

@rxwei: what do you think about this design?

jtkeeling commented:
Thanks for the comment, @Azoy 

As @dan-zheng describes, the design here is intended to make sure that random streams are reproducible, which is a useful property for experimentation. I agree with his proposal to reuse the `RandomNumberGenerator` protocol and it will simplify this PR, so thanks for the tip.

Azoy commented:
I understand about reproducibility, I was thinking along the lines of:

```swift
public protocol SeeadableRNG: RandomNumberGenerator {
  init(seed: [UInt8])
  init<T: BinaryInteger>(seed: T)
}
```

so that any custom generator tensorflow may need can take advantage of the stdlib provided facilities such as random ints/floats/bools, random element from collection, shuffling, etc.

natecook1000 commented:
Exactly — the uniform and normal distributions would be useful for all users of Swift, as would as a sendable ARC4Random implementation, so it would be great if they could participate in the `RandomNumberGenerator` design. Thanks!

jtkeeling commented:
The `SeedableRNG` protocol proposal makes a lot of sense. I'll modify this PR to do that.

I would be interested in adding to the `RandomNumberGenerator` design, if that makes sense, though I would prefer to get the changes into the TF branch sooner, as they are required for some other work.

jtkeeling committed:
Address comments on PR.

In particular, create SeedableRandomNumberGenerator that inherits from RandomNumberGenerator.

jtkeeling commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,141 @@
+  func generate(count: Int) -> [UInt8]
+}
+
+@_fixed_layout
```
I've added some comments in the latest commit.

dan-zheng commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -579,78 +579,26 @@ public extension Tensor where Scalar : Numeric {
   ///   - shape: The dimensions of the tensor.
-  ///   - state: The pseudorandom state in which the random numbers are being
-  ///     generated.
+  ///   - engine: The random engine which generates random bytes.
```
Rename `engine` in doc comment to `rng`, likewise below.

jtkeeling committed:
Fix comment as requested by dan-zheng

dan-zheng commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+// Distributions
+//===----------------------------------------------------------------------===//
+
+public class UniformIntegerDistribution<T: BinaryInteger> {
```
Does it make sense for the distribution classes to be `final`?
Or are there uses for subclassing them?

jtkeeling replied:
Yes, that makes sense. Will do.

jtkeeling committed:
Add final to distribution classes

dan-zheng reviewed (approved):
LGTM!
One note is: previously, the random uniform/normal distribution logic was defined inline in the Tensor random initializers. The distribution classes are a new addition and will need to go through evolution, if the plan is to upstream them.

natecook1000 commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+}
+
+@_fixed_layout
+public class UniformFloatingPointDistribution<T: BinaryFloatingPoint> {
```
If you add `T.RawSignificand : FixedWidthInteger` as a constraint here, you can use `T.random(in: a..<b, using: &rng)` in the `next()` method, which should be more efficient and accurate than integer division.

dan-zheng commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+// Distributions
+//===----------------------------------------------------------------------===//
+
+public class UniformIntegerDistribution<T: BinaryInteger> {
```
`UniformIntegerDistribution` does not have configurable low/high bounds.
Consider adding low/high bound functionality, or renaming it to `StandardUniformIntegerDistribution`?

dan-zheng commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+  public let b: T
+  private let uniformIntDist = UniformIntegerDistribution<UInt64>()
+
+  public init(a: T = 0, b: T = 1) {
```
Bikeshedding: could you rename `a` and `b` to `lowerBound` and `upperBound`?
The names `lowerBound` and `upperBound` are used in [`Range`](https://developer.apple.com/documentation/swift/range).

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+    return result
+  }
+
+  // Helper to access the state.
```
These helpers can be private.

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+// Distributions
+//===----------------------------------------------------------------------===//
+
+public final class UniformIntegerDistribution<T: BinaryInteger> {
```
Add `@_fixed_layout` here.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -662,14 +610,21 @@ public extension Tensor where Scalar : BinaryFloatingPoint {
   ///
   @inlinable @inline(__always)
-  init(randomUniform shape: TensorShape, state: RandomState? = nil) {
+  init(randomUniform shape: TensorShape, rng: RandomNumberGenerator? = nil) {
```
Change argument label from `rng:` to `generator:`.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -579,78 +579,26 @@ public extension Tensor where Scalar : Numeric {
   @inlinable @inline(__always)
-  init(randomStandardUniform shape: TensorShape, state: RandomState? = nil) {
+  init(randomStandardUniform shape: TensorShape,
+       rng: RandomNumberGenerator? = nil) {
```
Same here.

rxwei commented on stdlib/public/TensorFlow/Tensor.swift:
```diff
@@ -679,16 +634,23 @@ public extension Tensor where Scalar : BinaryFloatingPoint {
-    let boxMullerHelper = Tensor(randomUniform: shape, state: state)
-    let result = sqrt(-2 * log(uniform)) * cos(2 * .pi * boxMullerHelper)
-    self = result * stddev + mean
+       rng: RandomNumberGenerator? = nil) {
```
Same here.

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+
+@_fixed_layout
+public final class UniformFloatingPointDistribution<T: BinaryFloatingPoint> {
+  public let a: T
```
Rename `a` and `b` here to `lowerBound` and `upperBound`.

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+@_fixed_layout
+public struct ARC4RandomNumberGenerator: SeedableRandomNumberGenerator {
+  public static let global = ARC4RandomNumberGenerator(seed: UInt32(time(nil)))
+  var state: [UInt8] = Array(0 ... 255)
```
Remove whitespaces around `...`.

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+    _precondition(seed.count > 0, "Length of seed must be positive")
+    _precondition(seed.count <= 256, "Length of seed must be at most 256")
+    var j: UInt8 = 0
+    for i: UInt8 in 0 ... 255 {
```
Remove whitespaces around `...`.

rxwei commented on stdlib/public/TensorFlow/Random.swift:
```diff
@@ -0,0 +1,162 @@
+}
+
+@_fixed_layout
+public final class NormalFloatingPointDistribution<T: BinaryFloatingPoint> {
```
Since there's no "normal integer distribution". Would it make sense to rename this to just `NormalDistribution`?

rxwei reviewed (commented):
It's nice that there are standalone distribution data structures. Thanks for tackling this!

jtkeeling committed:
Address comments on PR

Mostly just renaming. Only large-ish change is adding the restriction: `where T.RawSignificand : FixedWidthInteger` which means we can use T.random(in: using:) to do some of the generation

jtkeeling commented:
Have addressed everyone's comments.

Only large-ish change is adding the restriction: `where T.RawSignificand : FixedWidthInteger` which means we can use T.random(in: using:) to do some of the generation as recommended by @natecook1000

jtkeeling committed:
fix comment

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please clean test tensorflow

rxwei commented:
@swift-ci please clean test tensorflow

rxwei commented:
@swift-ci please clean test tensorflow

rxwei commented:
@swift-ci please clean test tensorflow

jtkeeling committed:
Work around compiler bug by duplicating init() methods

rxwei commented:
@jtkeeling James, you can drop all `_TFHoistable` for now to unblock this patch and further progress. We'll develop a fix for `_TFHoistable`.

rxwei commented:
@jtkeeling James, do you still plan to merge this patch?

jtkeeling commented:
@rxwei Yes - will try to do today. Ran into some issues due to the _TFHoistable issues, but should be sorted now.

jtkeeling committed:
Rename _precondition to precondition

jtkeeling commented:
@rxwei I've synced this and tested it as follows:
* Running the tests
* Running the MNIST example with [randomNormal, randomUniform] x [optimization on, optimization off]

So I'm confident the issues with `_TFHoistable` are now fixed.

Could you do the merge for me? Thanks!

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
Thanks, James. Looks great!
